haystack/components/tools/tool_invoker.py

```diff
@@ -550,11 +550,21 @@ def run(
         # 2) Execute valid tool calls in parallel
         if tool_call_params:
             with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
+                # Helper: run a zero-arg callable within a captured contextvars.Context
```

Comment:
Could we follow the same design pattern as used for the fix in `run_async`? 

This would mean updating `_execute_single_tool_call` to handle the context copying. Or is there a reason why this doesn't work?

Reply:
_execute_single_tool_call runs inside the worker thread therefore copying context vars there would capture an empty/default context so the tracing parentage is lost again. contextvars must be copied in the caller thread before offloading to TP

Reply:
But let's see how to align run and run_async as much as possible so it is obvious. I had to do a bit of gymnastics here (see diff) due to test failures stemming from late-binding in closures and mypy so it looks scary - but it is not.

Reply:
(If it works) I am OK with the idea